### PR TITLE
fix: Token list should respect `hideZeroBalance` setting

### PR DESCRIPTION
**Pull request by gambinish**
## **Description**

If a token is not a native asset for a chain, and the balance is zero, it should not be rendered if the `hideZeroBalance` setting is toggled to true.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/29058?quickstart=1)

## **Related issues**

Fixes: erc20 and native tokens with zero balance rendering on asset list on 12.9 release.

## **Manual testing steps**

1. Add an erc20 token to any chain with zero balance, should display when imported by default
2. Toggle `hideZeroBalance` setting to true in general settings

> Respect the "hide zero balance" setting (when true):
          - Native tokens should always display with zero balance when on the current network filter.
          - Native tokens should not display with zero balance when on all networks filter
          - ERC20 tokens with zero balances should respect the setting on both the current and all networks.
   Respect the "hide zero balance" setting (when false):
          - Native tokens should always display with zero balance when on the current network filter.
          - Native tokens should always display with zero balance when on all networks filter
          - ERC20 tokens always display with zero balance on both the current and all networks filter.

## **Screenshots/Recordings**

https://github.com/user-attachments/assets/aaee7e23-ab5c-4216-89e5-dc11485200bc

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/app/assets/token-list/token-list.tsx [hunk 1]
```diff
@@ -147,14 +151,26 @@ export default function TokenList({
...
-            ...token,
-            balance,
-            tokenFiatAmount,
-            chainId,
-            string: String(balance),
-          });
+          if (hideZeroBalanceTokens) {
+            // only hide zero balance tokens if not native gas token
+            if (token.isNative || balance !== '0') {
+              tokensWithBalance.push({
```
Comment (bergeron) [inline ui/components/app/assets/token-list/token-list.tsx:157]: If this is the same push in both branches, could maybe be simpler with

```ts
if (!hideZeroBalanceTokens || token.isNative || balance !== '0') {
    // push it
}
```

Comment (gambinish) [reply ui/components/app/assets/token-list/token-list.tsx]: https://github.com/MetaMask/metamask-extension/pull/29058/commits/b6d235d135ff4952bcb65ad6d32c0a4e471f43eb

**Review by bergeron — DISMISSED** [review]

File: ui/components/app/assets/token-list/token-list.tsx [hunk 2]
```diff
@@ -147,14 +151,18 @@ export default function TokenList({
...
-            ...token,
-            balance,
-            tokenFiatAmount,
-            chainId,
-            string: String(balance),
-          });
+          // respect hide zero balance setting
+          // native tokens should still show zero balance regardless
+          // erc20s with zero balances should be hidden
+          if (!hideZeroBalanceTokens || token.isNative || balance !== '0') {
```
Comment (darkwing) [inline ui/components/app/assets/token-list/token-list.tsx:157]: Note: based on suggestion I think the `token.isNative` isn't a given and we'll need to add another condition here to only show a zero-balance native token if the "Current Network" is selected.  I think we should do that in a separate PR thought.

Comment (gambinish) [reply ui/components/app/assets/token-list/token-list.tsx:157]: It looks like @sahar-fehri added this condition added here: https://github.com/MetaMask/metamask-extension/pull/29073

Should we point and merge that PR into this one? They will need to be merged together eventually.

Comment (bergeron) [reply ui/components/app/assets/token-list/token-list.tsx:157]: I think they're simple enough to merge together.  They're changing the same if statement.

File: ui/components/app/assets/token-list/token-list.tsx [hunk 3]
```diff
@@ -109,7 +117,7 @@ export default function TokenList({
   const nativeBalances: Record<Hex, Hex> = useSelector(
     getSelectedAccountNativeTokenCachedBalanceByChainId,
   ) as Record<Hex, Hex>;
-
+  const isTestnet = useSelector(getIsTestnet);
```
Comment (bergeron) [inline ui/components/app/assets/token-list/token-list.tsx:120]: is this used?

**Review by darkwing — DISMISSED** [review]

**Review by sahar-fehri — DISMISSED** [review]

**Review by bergeron — APPROVED** [review]

**Review by vinnyhoward — APPROVED** [review]

**Review by darkwing — APPROVED** [review]
